FrontEnd/src/HomePage/MainPage.jsx

```diff
@@ -1,29 +1,88 @@
 import { useEffect, useState } from "react";
-import { fetchUserFeed } from "./RecommendationDB";
+import { fetchUserFeed, filterByCategory } from "./RecommendationDB";
 import CategoryList from "./CategoryList";
 import PostsPreview from "./PostsPreview";
 import PostFullDisplay from "./PostFullDisplay";
+import SearchBar from "../Search/SearchBar";
 import { useTheme } from "../UseContext";
+import { fetchAndUpdateUserLocation } from "../LocationUtil";
 import "./MainPage.css";
 
 function MainPage({ userUID, userData, socket }) {
   const [userFeed, setUserFeed] = useState([]);
   const [isLoading, setIsLoading] = useState(true);
   const [isPostDetailModalShown, setIsPostDetailModalShown] = useState(false);
+  const [filteredPosts, setFilteredPosts] = useState([]);
   const [selectedPost, setSelectedPost] = useState(null);
   const { theme } = useTheme();
 
-  const fetchFeed = async () => {
-    if (userUID !== null) {
-      const feedData = await fetchUserFeed(userUID);
-      setUserFeed(feedData);
-      setIsLoading(false);
-    }
-  };
   useEffect(() => {
-    fetchFeed();
+    const fetchDataAndUpdateLocalStorage = async () => {
+      try {
+        const storedUserFeed = localStorage.getItem("userFeed");
+        if (storedUserFeed !== null) {
+          const parsedUserFeed = JSON.parse(storedUserFeed);
+          setUserFeed(parsedUserFeed);
+          setFilteredPosts(parsedUserFeed);
+          setIsLoading(false);
+        }
+
+        if (userUID !== null || storedUserFeed === null) {
+          const feedData = await fetchUserFeed(userUID);
+          const feedDataWithDates = feedData.map((item) => ({
+            ...item,
+            createdAt: item.createdAt.toDate(),
+          }));
+          setUserFeed(feedDataWithDates);
+          setFilteredPosts(feedDataWithDates);
+          setIsLoading(false);
+          localStorage.setItem("userFeed", JSON.stringify(feedDataWithDates));
+        }
+
+        if (userUID !== null && userUID !== undefined) {
```

Comment:
This would cover both `null` and `undefined`:
```suggestion
        if (userUID != null) {
```